pkg/chunked/storage_linux.go

```diff
@@ -1734,3 +1754,63 @@ func validateChunkChecksum(chunk *internal.FileMetadata, root, path string, offs
 
 	return digester.Digest() == digest
 }
+
+func newStagedFileGetter(dirFD int, differOpts *graphdriver.DifferOptions, entriesWithOriginalNames []fileMetadata) (*stagedFileGetter, error) {
+	var pathMap map[string]string = nil
+	if differOpts != nil && differOpts.Format == graphdriver.DifferOutputFormatFlat {
+		pathMap = make(map[string]string, len(entriesWithOriginalNames))
+		// This needs to match cloneEntriesWithFlatNames
+		for _, e := range entriesWithOriginalNames {
+			if e.Type != TypeReg {
+				continue
+			}
+			if e.Digest == "" {
+				return nil, fmt.Errorf("missing digest for %q", e.Name)
+			}
+			digest, err := digest.Parse(e.Digest)
+			if err != nil {
+				return nil, err
+			}
+			d := digest.Encoded()
+			pathMap[filepath.Clean(e.Name)] = fmt.Sprintf("%s/%s", d[0:2], d[2:])
```

Comment:
The file naming logic should not be duplicated.

Reply:
Introduced `pkg/chunked/internal/composefs.RegularFilePathForValidatedDigest` for that.